circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,172 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+use num_traits::pow;
+use std::collections::HashMap;
+
+use crate::{params::RoundParams, poseidon::native::Poseidon};
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+/// MerkleTree structure
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// PhantomData for the params
+	_params: PhantomData<P>,
+	/// Variables with level and value to represent nodes
+	nodes: HashMap<usize, Vec<F>>,
+	/// Height of the tree
+	height: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Build a MerkleTree from given leaf nodes
+	fn build_tree(mut leaves: Vec<F>, height: usize) -> Self {
+		assert!(leaves.len() <= pow(2, height));
+		// 0th level is the leaf level and the max level is the root level
+		let mut nodes = HashMap::new();
+		// Assign zero to the leaf values if they are empty
+		for i in leaves.len()..pow(2, height) {
+			leaves.push(F::zero());
+		}
+		nodes.insert(0, leaves);
+
+		for level in 0..height {
+			let mut hash = Vec::new();
+			for i in 0..nodes[&level].len() {
+				if i % 2 == 0 {
+					let hasher: Poseidon<F, WIDTH, P> = Poseidon::new([
+						nodes[&level][i],
+						nodes[&level][i + 1],
+						F::zero(),
+						F::zero(),
+						F::zero(),
+					]);
+					hash.push(hasher.permute()[0]);
```

Comment:
We could do a little change here to avoid deep nesting and make it more visually clear.
1) we can replace `i % 2 == 0` with `i % 2 != 0 { continue; }`
2) Define Poseidon inputs before calling the `new` function:
```rust
let pos_inputs = [
  nodes[&level][i],
  nodes[&level][i + 1],
  F::zero(),
  F::zero(),
  F::zero(),
]
let hasher: Poseidon<F, WIDTH, P> = Poseidon::new(pos_inputs);
```